src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -195,4 +199,66 @@ public void close() {
         }
         return rdd;
     }
+
+
+    @Override
+    public Dataset<Row> getDataFrame(DynamicContext context) {
+        this._df = null;
+
+        if (this._child == null) {
+            // create initial RDD from expression
+            JavaRDD<Item> initialRdd = this.getNewRDDFromExpression(_expression);
+
+            // define a schema
+            String schemaString = _variableName;
+            List<StructField> fields = new ArrayList<>();
+            for (String fieldName : schemaString.split(" ")) {
+                StructField field = DataTypes.createStructField(fieldName, DataTypes.BinaryType, true);
+                fields.add(field);
+            }
+            StructType schema = DataTypes.createStructType(fields);
+
+            // convert initial RDD to row RDD
+            JavaRDD<Row> rowRDD = initialRdd.map(new ForClauseSerializeClosure());
```

Comment:
Perfect.

Reply:
For future reference (please don't worry about it just now): I *suspect* that passing Encoders.kryo(Item.class) to a DataFrame map() with a trivial, idempotent closure *may* directly take care of all the serialization for you, which could make a lot of the code dealing with kryo serialization superfluous.

https://spark.apache.org/docs/2.1.2/api/java/org/apache/spark/sql/Encoders.html#kryo(scala.reflect.ClassTag)